addons/info/src/index.js

```diff
@@ -1,74 +1,116 @@
+
+const addonOptions = {
+  globalOptions: defaultOptions,
+  localOptions: {},
+  isGlobalScope: true,
+
+  setOptions(options) {
+    if (this.isGlobalScope) {
+      this.globalOptions = {
+        ...this.globalOptions,
+        ...options,
+      };
+      return this.globalOptions;
+    }
+    this.localOptions = {
+      ...this.localOptions,
+      ...options,
+    };
+    return this.localOptions;
+  },
+
+  globalScope() {
+    this.isGlobalScope = true;
+    this.localOptions = {};
+  },
 
-const defaultMarksyConf = {
-  h1: H1,
-  h2: H2,
-  h3: H3,
-  h4: H4,
-  h5: H5,
-  h6: H6,
-  code: Code,
-  p: P,
-  a: A,
-  li: LI,
-  ul: UL,
+  localScope() {
+    this.isGlobalScope = false;
+    this.localOptions = {};
+  },
 };
 
+function catchLocalOptions(storyFn, context) {
+  addonOptions.localScope();
+  const story = storyFn(context);
+  const { localOptions, globalOptions } = addonOptions;
+  addonOptions.globalScope();
+  return { story, localOptions, globalOptions };
+}
+
+function sendToPanel(infoString) {
+  addons.getChannel().emit(EVENT_ID, {
+    infoString,
+  });
+}
+
 function addInfo(storyFn, context, infoOptions) {
+  // Options could be overridden by setInfoOptions during storyFn execution
+  const { story, globalOptions, localOptions } = catchLocalOptions(storyFn, context);
+
   const options = {
-    ...defaultOptions,
+    ...globalOptions,
     ...infoOptions,
+    ...localOptions,
+  };
+
+  const components = {
+    ...options.marksyConf,
+    ...options.components,
   };
 
   // props.propTables can only be either an array of components or null
   // propTables option is allowed to be set to 'false' (a boolean)
   // if the option is false, replace it with null to avoid react warnings
-  if (!options.propTables) {
-    options.propTables = null;
-  }
-
-  const marksyConf = { ...defaultMarksyConf };
-  if (options && options.marksyConf) {
-    Object.assign(marksyConf, options.marksyConf);
-  }
   const props = {
-    info: options.text,
+    summary: options.summary,
     context,
     showInline: Boolean(options.inline),
     showHeader: Boolean(options.header),
     showSource: Boolean(options.source),
-    propTables: options.propTables,
+    hideInfoButton: Boolean(!options.infoButton),
+    propTables: options.propTables || null,
     propTablesExclude: options.propTablesExclude,
-    styles: typeof options.styles === 'function' ? options.styles : s => s,
-    marksyConf,
+    styles: typeof options.styles === 'function' && options.styles,
+    components,
     maxPropObjectKeys: options.maxPropObjectKeys,
     maxPropArrayLength: options.maxPropArrayLength,
     maxPropsIntoLine: options.maxPropsIntoLine,
     maxPropStringLength: options.maxPropStringLength,
   };
-  return (
+
+  const infoContent = (
     <Story {...props}>
-      {storyFn(context)}
+      {story}
     </Story>
   );
+  if (options.sendToPanel) {
+    const infoString = ReactDOMServer.renderToString(infoContent);
```

Comment:
This renerToString nessesary here? Why? 

`ReactDOMServer` is for server usage right? Won't it blow the manager.bundle size out of the park?

Reply:
It's a good question! I'm not sure how can we get this without `ReactDOMServer`, but maybe you have ideas? More details why it is: https://github.com/storybooks/storybook/pull/1501#issuecomment-325218401